### PR TITLE
Smart linking MVP: Allow app id in URL to be an upstream app id

### PR DESCRIPTION
**Pull request by orangejenny**
PR into https://github.com/dimagi/commcare-hq/pull/30002

For cross-jurisdictional web apps workflows, we'll want to link from one app into the corresponding app in another domain. To accomplish this, if a smart link can't find the app in the URL, assume it's an upstream app id, and link to the downstream app in the current domain.

This will potentially fail if the same upstream app has multiple downstream copies in a single domain.

### CODE REVIEW

File: corehq/apps/linked_domain/applications.py [hunk 1]
```diff
@@ -67,15 +67,15 @@ def get_downstream_app_id(downstream_domain, upstream_app_id, use_upstream_app_i
...
+    return list(get_downstream_app_id_map(downstream_domain))
 
 
 @quickcache(
     vary_on=['downstream_domain', 'use_upstream_app_id'],
     skip_arg=lambda *args, **kwargs: settings.UNIT_TESTING,
     timeout=5 * 60
 )
-def _get_downstream_app_id_map(downstream_domain, use_upstream_app_id=False):
+def get_downstream_app_id_map(downstream_domain, use_upstream_app_id=False):
```
Comment (gherceg) [inline corehq/apps/linked_domain/applications.py:78]: I could have sworn I listened to your advice and made arguments like `use_upstream_app_id` default to True since that is what we want going forward. I must be thinking of a different method though. Not requesting any change as that is out of scope of this PR.

File: corehq/apps/cloudcare/views.py [hunk 2]
```diff
@@ -589,7 +590,11 @@ def _fail(error):
 
     build = _fetch_build(domain, request.couch_user.username, app_id)
     if not build:
-        return _fail(_("Could not find application."))
+        id_map = get_downstream_app_id_map(domain, use_upstream_app_id=True)
+        if app_id in id_map.keys():
+            build = _fetch_build(domain, request.couch_user.username, id_map[app_id])
```
Comment (gherceg) [inline corehq/apps/cloudcare/views.py:595]: I understand that we are assuming the app_id might reference an upstream app and therefore want to try fetching a downstream app in this domain based on that upstream app id, but I don't understand why that is a possibility. Could you elaborate on when this flow is triggered? 

And as a potential followup, if we are able to reflect that _why_ part of this in the code itself, maybe just a comment, that might be helpful in the future.

**Review by gherceg — APPROVED** [review]
Looks like a reasonable change. Left a non-blocking question and a comment.

File: corehq/apps/cloudcare/views.py (hunk 2 above)
Comment (orangejenny) [reply corehq/apps/cloudcare/views.py:595]: USH is doing a bunch of work to support cross-domain workflows, where the domains are all downstream from the same "development" domain. Check out the screenshot in [this PR](https://github.com/dimagi/commcare-hq/pull/30388): you're looking at patients that are spread across your current domain but also other domains. When you select a patient, if they're in a different domain, we want to use a smart link to send you into that domain at the same place in the "same" app, so the user has the same experience whether they select a patient in the current domain or in a different one.

**Comment by orangejenny** [comment]
@shubham1g5 I'm especially curious if this workflow is something that mobile could support. I'm guessing not, because I don't think the app itself knows that it's a downstream app or what its upstream app id is, so I don't think there would be a way to look up the proper corresponding app.

File: corehq/apps/linked_domain/applications.py (hunk 1 above)
Comment (orangejenny) [reply corehq/apps/linked_domain/applications.py]: 7c790dbe295d8bf083a9feea5bf33607a7e4f752

File: corehq/apps/cloudcare/views.py (hunk 2 above)
Comment (orangejenny) [reply corehq/apps/cloudcare/views.py]: a371816ebdf2f2c751c0370f74807f6394057326

File: corehq/apps/linked_domain/applications.py (hunk 1 above)
Comment (gherceg) [reply corehq/apps/linked_domain/applications.py]: Easy enough. Thank you!

File: corehq/apps/cloudcare/views.py [hunk 3]
```diff
@@ -589,7 +589,15 @@ def _fail(error):
 
     build = _fetch_build(domain, request.couch_user.username, app_id)
     if not build:
-        return _fail(_("Could not find application."))
+        # These links can be used for cross-domain web apps workflows, where a link jumps to the
+        # same screen but in another domain's corresponding app. This works if both the source and
+        # target apps are downstream apps that share an upstream app - the link references the upstream app.
+        from corehq.apps.linked_domain.applications import get_downstream_app_id_map
+        id_map = get_downstream_app_id_map(domain)
+        if app_id in id_map.keys():
```
Comment (snopoke) [inline corehq/apps/cloudcare/views.py:597]: nit:
```suggestion
        if app_id in id_map:
```

**Review by snopoke — APPROVED** [review]

File: corehq/apps/cloudcare/views.py [hunk 4]
```diff
@@ -589,7 +589,15 @@ def _fail(error):
 
     build = _fetch_build(domain, request.couch_user.username, app_id)
     if not build:
-        return _fail(_("Could not find application."))
+        # These links can be used for cross-domain web apps workflows, where a link jumps to the
```
Comment (shubham1g5) [inline corehq/apps/cloudcare/views.py:592]: @orangejenny I am having little trouble contextualizing this. In particular what are upstream and downstream apps and how they are different from source and traget apps here ?

File: corehq/apps/cloudcare/views.py (hunk 3 above)
Comment (orangejenny) [reply corehq/apps/cloudcare/views.py:597]: I always forget you can do that.

File: corehq/apps/cloudcare/views.py (hunk 4 above)
Comment (orangejenny) [reply corehq/apps/cloudcare/views.py:592]: I meant that "source" is the app that the user is currently looking at in web apps, while "target" is the app that they're going to. Both the source and target apps are linked/downstream apps of the same upstream app.

So you're in county A, looking at patients from counties A, B, C, and D. If you click on a patient in B, C, or D, you need a smart link. I'd like to use one smart link template for all of those patients, something like `/a/{DOMAIN}/app/v1/{APP_ID}/endpoint_id`. Web apps / formplayer can replace the domain in the URL, since the case will know what domain it belongs to (thanks to https://github.com/dimagi/commcare-hq/pull/30388), but it doesn't know what app id to use. I'm using the upstream app id in the URL and then having HQ look up the actual app to redirect to.

Comment (shubham1g5) [reply corehq/apps/cloudcare/views.py:592]: makes sense, thanks for the detailed explanation.

**Comment by shubham1g5** [comment]
>  I'm especially curious if this workflow is something that mobile could support. 

@orangejenny  Yeah this seems tricky to support on mobile for the reasons you have mentioned. The best we can do on mobile is try to see if the same endpoint exists for any of the installed apps on phone if we don't find the endpoint for the given app id or if the corresponding app with the given app id is not installed on the phone. Though it would be quite process heavy on mobile to loop through all the installed apps and look for the right endpoint definition in them.

**Comment by orangejenny** [comment]
@shubham1g5 Cool, that makes sense.
